### PR TITLE
Apollo batching

### PR DESCRIPTION
**Pull request by AlexCLeduc**
This involved moving from express-graphql to apollo-server. 

- [x] Try getting this to work with our special GET requests (we'll probably have to move back to GET + body, hopefully we can get that to cache properly)
- [x] See if batching works in dev-links
- [x] ~~cloudflare considerations?~~ (we weren't cloudflare caching anything before)
- [x] graphiql considerations (apollo-server seems to replace graphiql with apollo playground)
    - infobase's `#graphiql` route still works, but since introspection is turned off in prod, it won't support docs or autocomplete unless you're developing locally
- [x] turn `validationRules: [depthLimit(15)]` security back on using apollo-server options 
- [x]  Cross browser testing
    - [x] IE11 and iOS both make requests and get responses properly
    - [x] unfortunately we need to add a `String.prototype.replaceAll` polyfill for IE11 

Screen Shot 2021-04-05 at 12 23 30 PM

### CODE REVIEW

**Comment by AlexCLeduc** [comment]
Got it to work with GET requests, and it seems to cache locally

Screen Shot 2021-04-08 at 9 33 29 PM
Screen Shot 2021-04-08 at 9 34 15 PM

File: client/src/graphql_utils/graphql_utils.js
Comment (AlexCLeduc) [inline client/src/graphql_utils/graphql_utils.js:71]: Note that I stopped encoding it to header that holds the query. It's less annoying to debug, and as far as I can tell HTTP headers don't need to be encoded. 

That being said, HTTP headers should be ASCII, and might not support UTF8. This might cause an issue when we start using the API for search queries, and someone searches for something with accents. If this causes a bug, we can just `_.deburr` queries. All the searching should ignore diacritics/accents anyway.

File: client/src/models/covid/populate.js
Comment (AlexCLeduc) [inline client/src/models/covid/populate.js:26]: Apollo server doesn't cast integers to strings the way express-graphql does, so I did that manually. 

We could alternatively make the department queries in the schema take integer arguments 🤷

File: server/src/app.js
Comment (AlexCLeduc) [inline server/src/app.js:59]: Since a request can now contain multiple queries, we log each query. Part of the log objects is a batch_hash, which can be used to identify which query logs were part of the same request

File: client/src/graphql_utils/graphql_utils.js
Comment (Stephen-ONeil) [inline client/src/graphql_utils/graphql_utils.js:58]: This comment no longer applies

File: client/src/graphql_utils/graphql_utils.js
Comment (Stephen-ONeil) [reply client/src/graphql_utils/graphql_utils.js:71]: I believe this, or a similar post, was what informed my previous decision to compress: https://stackoverflow.com/questions/3326210/can-http-headers-be-too-big-for-browsers#:~:text=My%20Conclusion%3A%20If%20you%20want,256KB%20for%20all%20headers%20together
  
Don't know if Firefox still has the same 10KB per-header limit. That's still possibly up to 10k characters of query though, so it _sounds_ like we're totally safe even with batching growing query sizes? I'll try and dig up the old PR where I added this just to check whether I was adding it out of extreme caution, or if I'd actually had issues in the wild.

I do agree it'd be nice to stop it, it really is a huge pain when debugging.

File: client/src/models/covid/populate.js
Comment (Stephen-ONeil) [reply client/src/models/covid/populate.js:26]: Ouch. Yeah, the inconsistency in id types between int and string bugs me (and creates bugs) a lot. Org ids should probably just always be strings, doesn't really matter that they happen to only have int values. Other subjects pretty much all have non-int ids so it's only the org subjects that have the ambiguity headache.

File: server/package.json
Comment (Stephen-ONeil) [inline server/package.json:21]: We can drop the express-graphql and compression dependencies now, right? Any other packages we've stopped using here?

File: server/src/app.js
Comment (Stephen-ONeil) [inline server/src/app.js:26]: Hmm, don't even remember what this was for. Maybe introspection queries got too big for the express defaults? Probably should have only been on in dev if that was the case, oops.

**Review by Stephen-ONeil — COMMENTED** [review]
Awesome! Mostly minor comments here. Going to test out in a bunch of browsers and try and sniff out any gotchas before approving.

**Comment by Stephen-ONeil** [comment]
Follow up on what I asked about in the dev meeting, can confirm that (as expected) Apollo still de-dupes queries in batches and skips queries that are already in-flight/cached. 

Well, hard to say for certain about in-flight requests, timing being what it is, but pretty confident about that either way. No reason for Apollo to lose the ability to be smart about that just because batching is on. 

If you want to look at that your self, a good place to do it is in the covid infographic (with `COVID_EXPENDITUES_FLAG` changed to true in covid_config.js). Both the covid authorities and covid expenditures panels make a query for `query_gov_covid_summaries` on mount, but it only shows up once in all the batched queries.

File: client/src/graphql_utils/graphql_utils.js
Comment (Stephen-ONeil) [reply client/src/graphql_utils/graphql_utils.js:71]: Original API GET requests PR: https://github.com/TBS-EACPD/infobase-archive/pull/753

Didn't leave much explanation (either of the problem or solution), possible I never figured out what was actually wrong. Looks like I ruled out the per-header limit being the issue at least. Sounds like certain longer queries were somehow throwing CORs errors, and I don't mention any particular browser so it must have been a fully server side issue. Don't see it happening anywhere now though 🤷 .

**Review by Stephen-ONeil — APPROVED** [review]
Everything seems solid to me! Pre-approved, pending
  - [x] any action you want to take from the minor comments above
  - [x] polyfilling `replaceAll`
  
If you're busy on Titan, I can wrap this up for you 👍

File: server/package.json
Comment (AlexCLeduc) [reply server/package.json:21]: Good point, forgot to remove express-graphql. 

We're still using compression to gzip responses, which are going to become even more redundant with batching.

File: server/src/app.js
Comment (AlexCLeduc) [reply server/src/app.js]: nice find, got rid of it

**Review by ktw1016 — COMMENTED** [review]
I looked through and mostly seems good! Would just need to clean up the express graphql stuff before merging. I tried testing it out with covid and results but difference was pretty minimal as they only make 2-3 queries. I'll experiment it with service inventory queries and let you know once it merges

**Pull request merged**
